### PR TITLE
As a user I should be able to see all my conversations

### PR DESCRIPTION
**Pull request by keinydev**
[Trello card here](https://trello.com/c/PYcxvv4Z/10-3-as-a-user-i-should-be-able-to-see-all-my-conversations)

### CODE REVIEW

File: app/controllers/api/v1/conversations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+module Api
+  module V1
+    class ConversationsController < ApplicationController
+      
+      before_action :authenticate_user!
+
+      def index
+        @conversations = Conversation.by_user(current_user.id)
```
Comment (enzofab91) [inline app/controllers/api/v1/conversations_controller.rb:8]:
```suggestion
        @conversations = current_user.conversations
```

File: app/controllers/api/v1/targets_controller.rb [hunk 2]
```diff
@@ -31,6 +43,22 @@ def target_params
       def target
         @target = Target.find(params[:id])
       end
+
+      def search_match(radius, latitude, longitude, topic_id)
```
Comment (enzofab91) [inline app/controllers/api/v1/targets_controller.rb:47]: This is a private method, isn't it?

File: app/controllers/api/v1/targets_controller.rb [hunk 3]
```diff
@@ -11,7 +11,19 @@ def index
       def create
         @target = current_user.targets.build(target_params)
         if @target.save
-          @target
+          match = search_match(@target.radius, @target.latitude, @target.longitude, @target.topic_id)
+          matched_user = {}
+          match_conversation = {}
+          
+          if match.size > 0
```
Comment (enzofab91) [inline app/controllers/api/v1/targets_controller.rb:18]: I think you should iterate over match collection. The search could find multiple matches among users for the same topic.

File: spec/requests/api/v1/conversations/index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+  subject(:get_conversations) { get api_v1_conversations_path, headers: auth_headers, as: :json }
+
+  describe 'GET Index' do
+    context 'when params are valid' do
+      before do
+        get_conversations
+      end
+      
+      it 'returns status 200' do
```
Comment (enzofab91) [inline spec/requests/api/v1/conversations/index_spec.rb:36]:
```suggestion
      it 'returns a successful response' do
```

**Review by enzofab91 — COMMENTED** [review]
Good work!

Left some comments.

File: app/controllers/api/v1/targets_controller.rb (hunk 3 above)
Comment (keinydev) [reply app/controllers/api/v1/targets_controller.rb:18]: It is possible, but the requirement says one match. Only one.

Comment (enzofab91) [reply app/controllers/api/v1/targets_controller.rb:18]: Yes, only one conversation between **_two_** users, but you can have a conversation with each user in the system. This function find every potencial user to create a match (conversation) with.

File: app/controllers/api/v1/targets_controller.rb (hunk 2 above)
Comment (keinydev) [reply app/controllers/api/v1/targets_controller.rb:47]: Yes

File: app/controllers/api/v1/targets_controller.rb (hunk 3 above)
Comment (keinydev) [reply app/controllers/api/v1/targets_controller.rb:18]: Understood and solved!

**Review by enzofab91 — APPROVED** [review]
👏

File: app/controllers/api/v1/targets_controller.rb [hunk 5]
```diff
@@ -31,6 +47,22 @@ def target_params
...
+      def search_match(radius, latitude, longitude, topic_id)
+        Target.within(
+          radius,
+          :units => :miles,
+          :origin => [ latitude, longitude ]
+        ).where(topic_id: topic_id).where.not(user_id: current_user.id)
+      end
+
+      def set_conversation(user_a, user_b)
+        if Conversation.between(user_a, user_b).present?
```
Comment (santir489) [inline app/controllers/api/v1/targets_controller.rb:60]: Maybe you can use `first_or_create`

https://apidock.com/rails/ActiveRecord/Relation/first_or_create

File: app/models/conversation.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+#  updated_at :datetime         not null
+#
+class Conversation < ApplicationRecord
+  belongs_to :user_a, foreign_key: :user_a_id, class_name: "User"
+  belongs_to :user_b, foreign_key: :user_b_id, class_name: "User"
+
+  validates :user_a, uniqueness: {scope: :user_b}
+
+  scope :between, -> (user_a_id, user_b_id) do
+    where("(conversations.user_a_id = ? AND conversations.user_b_id = ?) OR (conversations.user_a_id = ? AND conversations.user_b_id = ?)", user_a_id, user_b_id, user_b_id, user_a_id)
```
Comment (santir489) [inline app/models/conversation.rb:18]: you can name params, so you do not have to repeat them. For example, you are setting user_a_id twice

File: spec/requests/api/v1/conversations/index_spec.rb [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      latitude: 22.3129125,
+      radius: 5,
+      topic_id: topic.id
+    }    
+  }
+  end
+
+  before do
+    post api_v1_targets_path, params: target_params, headers: auth_headers, as: :json 
+  end
```
Comment (santir489) [inline spec/requests/api/v1/conversations/index_spec.rb:26]: What you are testing is getting conversations... so there is no need to create this using the controller. Try to use factories to create the test context

File: app/controllers/api/v1/targets_controller.rb [hunk 8]
```diff
@@ -11,7 +11,23 @@ def index
       def create
         @target = current_user.targets.build(target_params)
         if @target.save
-          @target
+          match = search_match(@target.radius, @target.latitude, @target.longitude, @target.topic_id)
+          matched_user = []
+          match_conversation = []
+          
+          if match.size > 0
```
Comment (santir489) [inline app/controllers/api/v1/targets_controller.rb:18]: This create methods is getting too long. There is a good practice that says: keep controller methods skinny. You can move this logic to a create target service.

This comment is not mandatory but to keep it in mind. Feel free to ignore me 😁

**Review by santir489 — COMMENTED** [review]
Nice work! Left some comments

File: spec/requests/api/v1/conversations/index_spec.rb (hunk 7 above)
Comment (keinydev) [reply spec/requests/api/v1/conversations/index_spec.rb:26]: Ok, I have done some changes

**Review by santir489 — APPROVED** [review]

**Pull request closed without merging**